### PR TITLE
Fix split_batch bug with empty generation_kwargs

### PR DESCRIPTION
**Pull request by maxisawesome**
# What does this PR do?

Split batch on an empty dict doesn't split as we'd want it to, so avoid this problem by not setting generation_kwargs instead of setting it to {} 
# What issue(s) does this change relate to?
`mpt-eval-tjHpGA` has an example of the bug occurring.
`test-jeopardy-cxh2Vi` reproduction of the bug. **This required the same number of gpus + device_eval_batch_size.**
`test-jeopardy-wYdXWy` has debugging info printed.
`test-jeopardy-nHufZD` is a successful run

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [x] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Comment by bmosaicml** [comment]
The errors appear to be different between the two runs:

The mpt-eval bug:
```
Traceback (most recent call last):
  File "/llm-foundry/scripts/eval/eval.py", line 491, in <module>
    main(cfg)
  File "/llm-foundry/scripts/eval/eval.py", line 337, in main
    eval_gauntlet_df) = evaluate_model(
  File "/llm-foundry/scripts/eval/eval.py", line 225, in evaluate_model
    trainer.eval(eval_dataloader=evaluators)
  File "/usr/lib/python3/dist-packages/composer/trainer/trainer.py", line 2841, in eval
    self._eval_loop(
  File "/usr/lib/python3/dist-packages/composer/trainer/trainer.py", line 2960, in _eval_loop
    self.state.batch = data_spec.split_batch(self.state.batch,
  File "/usr/lib/python3/dist-packages/composer/core/data_spec.py", line 91, in _default_split_batch
    return _split_mapping(batch, microbatch_size)
  File "/usr/lib/python3/dist-packages/composer/core/data_spec.py", line 63, in _split_mapping
    return [{k: v[idx] for k, v in chunked.items()} for idx in range(num_chunks)]
  File "/usr/lib/python3/dist-packages/composer/core/data_spec.py", line 63, in <listcomp>
    return [{k: v[idx] for k, v in chunked.items()} for idx in range(num_chunks)]
  File "/usr/lib/python3/dist-packages/composer/core/data_spec.py", line 63, in <dictcomp>
    return [{k: v[idx] for k, v in chunked.items()} for idx in range(num_chunks)]
IndexError: list index out of range
```

The test-jeopardy bug

```
2024-01-27 23:12:59,454: rank7[1075][MainThread]: INFO: composer.core.data_spec: {'input_ids': (tensor([[ 2300,   947, 48697,  ...,     0,     0,     0],
        [ 2300,   947, 48697,  ...,     0,     0,     0],
        [ 2300,   947, 48697,  ...,     0,     0,     0]], device='cuda:7'), tensor([[ 2300,   947, 48697,  ...,     0,     0,     0]], device='cuda:7')), 'continuation_indices': [[tensor([133, 134, 135, 136, 137, 138, 139], device='cuda:7'), tensor([117, 118], device='cuda:7'), tensor([122, 123, 124, 125, 126], device='cuda:7')], [tensor([124, 125], device='cuda:7')]], 'labels': (tensor([[ 2300,   947, 48697,  ...,     0,     0,     0],
        [ 2300,   947, 48697,  ...,     0,     0,     0],
        [ 2300,   947, 48697,  ...,     0,     0,     0]], device='cuda:7'), tensor([[ 2300,   947, 48697,  ...,     0,     0,     0]], device='cuda:7')), 'generation_kwargs': [{}], 'attention_mask': (tensor([[ True,  True,  True,  ..., False, False, False],
        [ True,  True,  True,  ..., False, False, False],
        [ True,  True,  True,  ..., False, False, False]], device='cuda:7'), tensor([[ True,  True,  True,  ..., False, False, False]], device='cuda:7')), 'mode': ['icl_task', 'icl_task']}
Traceback (most recent call last):
  File "/llm-foundry/scripts/eval/eval.py", line 505, in <module>
    main(cfg)
  File "/llm-foundry/scripts/eval/eval.py", line 349, in main
    eval_gauntlet_df) = evaluate_model(
  File "/llm-foundry/scripts/eval/eval.py", line 228, in evaluate_model
    trainer.eval(eval_dataloader=evaluators,
  File "/composer/composer/trainer/trainer.py", line 2841, in eval
    self._eval_loop(
  File "/composer/composer/trainer/trainer.py", line 2960, in _eval_loop
    self.state.batch = data_spec.split_batch(self.state.batch,
TypeError: 'NoneType' object is not subscriptable
```

**Comment by maxisawesome** [comment]
When debugging with extra log statements, I bracket the return statement in a try/except. The function thus returns None instead of the dictionary, and will error on `TypeError: 'NoneType' object is not subscriptable`. This only happens in the run with printed debugging info because I do not include the debugging logging in the final PR.

File: tests/datasets/test_in_context_learning_datasets.py [hunk 1]
```diff
@@ -255,7 +255,7 @@ def test_update_generation_kwargs_no_kwargs(tiny_gpt2_tokenizer, tmp_path):
                                   destination_path=str(tmp_path / 'test_dataset_lm_juggernaut.jsonl'),
                                   hf_loading_vars=hf_loading_vars,
                                   hf_parsing_map=hf_parsing_map)
-    assert not dl.base_batch['generation_kwargs']
+    assert not dl.base_batch.get('generation_kwargs')
```
Comment (dakinggg) [inline tests/datasets/test_in_context_learning_datasets.py:258]: this should be `assert not 'generation_kwargs' in dl.base_batch`, right?

File: composer/datasets/in_context_learning_evaluation.py [hunk 2]
```diff
@@ -349,7 +350,7 @@ def __len__(self) -> int:
     def get_num_samples_in_batch(self, batch: Dict) -> int:
         return batch['input_ids'].shape[0]
 
-    def update_generation_kwargs(self, generation_kwargs: Dict) -> None:
+    def update_generation_kwargs(self, generation_kwargs: Optional[Dict]) -> None:
```
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py:353]: remove `Optional` type

**Review by dakinggg — APPROVED** [review]

File: composer/datasets/in_context_learning_evaluation.py (hunk 2 above)
Comment (maxisawesome) [reply composer/datasets/in_context_learning_evaluation.py]: good point

**Pull request merged**